docs: Coverage Converter design document (#693)
platform-platform/monorepo: metrics/coverage_converter/docs/01_coverage_converter_design.md

![Interfaces diagram](http://www.plantuml.com/plantuml/proxy?cache=no&fmt=svg&src=https://raw.githubusercontent.com/platform-platform/monorepo/codecov_converter_design/metrics/coverage_converter/docs/diagrams/coverage_interfaces_diagram.puml)

As we can see above, the `CoverageConverterCommand` uses the `CoverageConverter` to convert the specific coverage report to the `CoverageModel` class that represents the `CI integrations` coverage report.

Let's consider the activity diagram of the coverage conversion process: 

![Coverage conversion diagram](http://www.plantuml.com/plantuml/proxy?cache=no&fmt=svg&src=https://github.com/platform-platform/monorepo/raw/codecov_converter_design/metrics/coverage_converter/docs/diagrams/coverage_conversion_activity.puml)

To run the command line application, we should have a `CommandRunner` class. Let's name it `CoverageConverterRunner`. This class will extend `CommandRunner`. It will be the place where we'll add our specific `CoverageConverterCommand`s.

Also, we should be able to pass the following arguments to the Coverage Converter tool: 

- input - the file from where we'll read the specific coverage report;

> - output - the file to which we'll write the formatted coverage report;

Reviewer: As we have a specific file name used for coverage artifact should we really provide the ability to specify output? Or we want to make sure that if this parameter is not provided - the default name will be used instead?

>     - command
>     - runner
>     - converter
>   - specific_format
>     - command
>     - model
>     - converter

## Creating a new specific format converter

Once we have a `CoverageConverterCommand` abstract class, the `CoverageConverter` interface, and the `CoverageConverterRunner`,  we can create the first specific coverage parser. To implement it, we should follow the next steps: 

1. Create a specific implementation of the `CoverageConverter`.
2. Create a specific command extending the `CoverageConverterCommand` that will use the specific `CoverageConverter` created in the previous step.
3. Add the previously created `CoverageConverterCommand` to the `CoverageConverterRunner` using the `addCommand` method.

Currently, the Coverage Converter tool should support the following coverage report formats: 

- LCOV
- Istanbul

## Summary

Finally, let's consider the class diagram that provides information about all classes and their relationships.

![Coverage class diagram](http://www.plantuml.com/plantuml/proxy?cache=no&fmt=svg&src=https://github.com/platform-platform/monorepo/raw/codecov_converter_design/metrics/coverage_converter/docs/diagrams/coverage_class_diagram.puml)

> So, the application takes the arguments, parses them using the `CoverageArgumentsParser`, and passes these arguments to the specific converter command. The converter command, in its turn, gets the input file path, converts it to the CI integrations coverage format, and writes it to the output file.

Reviewer: Should we somehow cover what will happen if File not found? Or if we can't write output file (no disk space, etc..)?

Let's consider the sequence diagram of these processes taking the `specific` format as an example:

![Coverage sequence diagram](http://www.plantuml.com/plantuml/proxy?cache=no&fmt=svg&src=https://github.com/platform-platform/monorepo/raw/codecov_converter_design/metrics/coverage_converter/docs/diagrams/coverage_sequence_diagram.puml)

# Dependencies

> What is the project blocked on?

This project has no blockers.

> What will be impacted by the project?

The mechanism of loading the coverage report in the CI integrations will be impacted.

# Testing

> How will the project be tested?

The project will be tested using unit tests.

# Alternatives Considered

> Summarize alternative designs (pros & cons)

No alternatives considered.

# Timeline

> Document milestones and deadlines.

DONE:

>   -

Reviewer: ```suggestion
  - Coverage Converter design.
```

NEXT:

  - Create main classes and interfaces.
  - Implement Coverage converter for the `lcov` format.
  - Implement Coverage converter for the `istanbul` format.
  
# Results

> What was the outcome of the project?

> Work in progress.

Reviewer: ```suggestion
Coverage converter tool design complete and ready for implementation.
```

# Coverage Converter design

> Summary of the proposed change

An explanation of the Coverage Converter tool architecture design.

# References

> Link to supporting documentation, GitHub tickets, etc.

- [LCOV Code Coverage](https://wiki.documentfoundation.org/Development/Lcov)
- [Istanbul](https://istanbul.js.org/)

# Motivation

> What problem is this project solving?

Since we have a `CI integrations` tool that stands for importing the CI builds to the Metrics application, we need to sure that the user can use this tool with any format of the coverage reports. To do so, we should create a separate, specific for `CI integrations`, coverage report format, and the tool that will convert the coverage reports with different formats into this format - the Coverage Converter tool. To make the Coverage Converter tool architecture clean and understandable, we need to create a document that will explain the main components and principles of the Coverage tool.

# Goals

> Identify success metrics and measurable goals.

- Explain the purpose of the Coverage Converter tool.
- Explain the main components of the Coverage Converter tool.

> - Explain adding the new format converter.

Reviewer: ```suggestion
```
We probably want to move this to `CI Integrations` documentation.

# Non-Goals

> Identify what's not in scope.

The document does not explain and shows the implementation details.

# Design

> Explain and diagram the technical design

## Main interfaces and classes

> The main purpose of this tool is to convert the given coverage report to the `CI integrations` readable one. To do so, we should create the following interfaces and abstract classes: 

Reviewer: ```suggestion
The main purpose of this tool is to convert the given supported coverage report to the `CI integrations` format. To do so, we should create the following interfaces and abstract classes: 
```
And maybe we should include a link to `CI Integrations` format?

Also, we should be able to pass the following arguments to the Coverage Converter tool: 

- input - the file from where we'll read the specific coverage report;
- output - the file to which we'll write the formatted coverage report;

To be able to pass arguments to our application we should create a `CoverageArgumentsParser` that will take a `List<String>` as arguments and return the `CoverageConverterArguments` instance that will contain all given arguments.

## Package structure

Once we've finished with all main classes and interfaces, let's consider the package structure of the `Coverage Converter`: 

> lib
>   - arguments
>     - model
>     - parser
>   - common 
>     - command
>     - runner
>     - converter
>   - specific_format
>     - command
>     - model
>     - converter

## Structure of the Coverage coverage report.

Since we want to unify the coverage reports for the CI integrations, we should create a specific for CI integrations coverage report format. Currently, this report will contain only `total coverage percent`. Let's consider the JSON summary structure: 

```
{
  pct: 0.6
}
```

> So, the JSON will contain only `pct` value that will represent the total coverage percent. Let's consider the class diagram of the coverage structure: 

Reviewer: Should we add more details like data type, range, precision, etc..? Is it compatible with Istanbul format? Can it have other fields?
